enclave_build/src/cache.rs

```diff
@@ -0,0 +1,734 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// S&PDX-License-Identifier: Apache-2.0
+
+use std::{
+    collections::HashMap,
+    fs::{self, File, OpenOptions},
+    io::{Error, ErrorKind, Read, Write},
+    path::{Path, PathBuf},
+};
+
+use sha2::Digest;
+
+use crate::{
+    image::{self, ImageDetails},
+    EnclaveBuildError, Result,
+};
+
+use oci_distribution::client::ImageData;
+
+/// Root folder for the cache.
+pub const CACHE_ROOT_FOLDER: &str = "XDG_DATA_HOME";
+/// Name of the cache index file which stores the (image URI <-> image hash) mappings.
+pub const CACHE_INDEX_FILE_NAME: &str = "index.json";
+/// The name of the folder used by the cache to store the image layers.
+pub const CACHE_LAYERS_FOLDER_NAME: &str = "layers";
+/// The name of the image config file from the cache.
+pub const CACHE_CONFIG_FILE_NAME: &str = "config.json";
+/// The name of the image manifest file from the cache.
+pub const CACHE_MANIFEST_FILE_NAME: &str = "manifest.json";
+
+/// Struct which provides operations with the local cache.
+///
+/// The index.json file is located in the cache root folder and keeps track of images stored in cache.
+///
+/// The cache structure is:
+///
+/// {CACHE_ROOT_PATH}/index.json\
+/// {CACHE_ROOT_PATH}/image_cache_folder1\
+/// {CACHE_ROOT_PATH}/image_cache_folder2\
+/// etc.
+///
+/// An image cache folder contains:
+///
+/// {IMAGE_FOLDER_PATH}/config.json - the image configuration, stored as a JSON String.\
+/// {IMAGE_FOLDER_PATH}/manifest.json - the image manifest, stored as a JSON String.\
+/// {IMAGE_FOLDER_PATH}/layers - folder containing all layers, each in a separate gzip compressed tar file.
+#[derive(Clone)]
+pub struct CacheManager {
+    /// The root folder of the cache
+    root_path: PathBuf,
+
+    /// A map storing the cached images, with the map entry format being (image_reference, image_hash)
+    cached_images: HashMap<String, String>,
+}
+
+impl CacheManager {
+    /// Creates a new CacheManager instance and returns it. As argument, a path to the root folder
+    /// of the cache should be provided.
+    ///
+    /// Apart from that, the function also creates (if not already created) all folders from the path
+    /// specified as argument.
+    ///
+    /// If an index.json file exists at the path, it loads the file's contents into the 'cached_images'
+    /// field. If not, a new empty index.json file is created at that path.
+    pub fn new<P: AsRef<Path>>(root_path: P) -> Result<Self> {
+        // Create all missing folders, if not already created
+        fs::create_dir_all(&root_path).map_err(EnclaveBuildError::CacheInitError)?;
+
+        let mut contents = String::new();
+
+        // Try to open the index.json file and read the contents.
```

Comment:
I find it pretty confusing to define a constant for some string, and then in comments use that particular string. We could instead say "try to open the cache index file".